highfidelity/hifi
# CR #21180 Camera control app

This PR implements a Camera Movement App that lets you configure easing parameters to smooth-out Avatar movements and turning, helping to avoid certain kinds of jitter when filming in an Avatar-as-a-Camera scenario.  Worklist item: https://worklist.net/21180

<img src='https://user-images.githubusercontent.com/94753/27406984-b7b632ee-56a4-11e7-857c-631a4aaf4c02.png'  />

* *Note: this app is intended only for Desktop mode -- previous discussions/specs indicated that some kind of warning should be displayed if someone attempts use in HMD (TBD?)*
*  *the "Avatars snap look at camera" and "Enable minimal mouse cursor" checkboxes depend on PR#10625 -- so are only fully-testable if running that build.  The rest of the app / settings can be tested using the most recent beta release.*

### Testing
* Load the `app-camera-move.js` Client script from the `unpublishedScripts/marketplace/camera-move` staging area; here's a rawgit link to use:
[unpublishedScripts/marketplace/camera-move/app-camera-move.js](https://cdn.rawgit.com/humbletim/hifi/c9874a7/unpublishedScripts/marketplace/camera-move/app-camera-move.js)
* Verify the new "Cam Drive" button appears in the toolbar.
* Press <kbd>space</kbd> to activate Camera Move Mode.
* Verify the app icon becomes "active".
* Verify mouse looking and keyboard movements now have smoothing applied.
* Experiment with various slider positions and values -- verifying that in general the changes get applied to avatar movements.
* Observe the current state of the "Enable snap turn in HMD" checkbox.
    * Open "Settings > Avatar" and change the "Snap turn when in HMD" setting.
    * Back in the Camera Movement App, verify the "Enable snap turn in HMD" checkbox reflects the new state.
* Press <kbd>space</kbd> to de-active Camera Move Mode.
* Verify avatar movements/turning take place normally again (ie: without any smoothing applied).
